### PR TITLE
Turn TextIOWrapper(buffer) into a protocol

### PR DESCRIPTION
**Pull request by srittau**
Closes: #11418

### CODE REVIEW

**Comment by AlexWaygood** [comment]
Why do we need all those empty `__init__.py` files in the `test_cases` directory? They appear to be confusing our `regr_test.py` script a bit

**Comment by srittau** [comment]
> Why do we need all those empty `__init__.py` files in the `test_cases` directory? They appear to be causing some confusion for our `regr_test.py` script

Because otherwise mypy gets confused about `check_io.pyi` and `typing/check_io.pyi`. But renaming the latter should help as well.

**Comment by AlexWaygood** [comment]
> > Why do we need all those empty `__init__.py` files in the `test_cases` directory? They appear to be causing some confusion for our `regr_test.py` script
> 
> Because otherwise mypy gets confused about `check_io.pyi` and `typing/check_io.pyi`. But renaming the latter should help as well.

Ugh. Yeah, renaming seems the easiest option there for now... thanks!

File: stdlib/io.pyi [hunk 1]
```diff
@@ -146,10 +149,33 @@ class TextIOBase(IOBase):
     def readlines(self, __hint: int = -1) -> list[str]: ...  # type: ignore[override]
     def read(self, __size: int | None = ...) -> str: ...
 
+@type_check_only
+class _WrappedBuffer(Protocol):
+    @property
+    def name(self) -> str: ...
+    @property
+    def closed(self) -> bool: ...
+    def read(self, size: int = ...) -> bytes: ...
```
Comment (AlexWaygood) [inline stdlib/io.pyi:158]: Are there any instances of `size` being passed as a keyword argument? If not, the parameter in the protocol should be pos-only:

```suggestion
    def read(self, size: int = ..., /) -> bytes: ...
```

File: stdlib/io.pyi [hunk 2]
```diff
@@ -146,10 +149,33 @@ class TextIOBase(IOBase):
...
+    # Optional: def read1(size: int, /) -> bytes: ...
```
Comment (AlexWaygood) [inline stdlib/io.pyi:159]:
```suggestion
    # Optional: def read1(self, size: int, /) -> bytes: ...
```

File: stdlib/io.pyi [hunk 3]
```diff
@@ -94,7 +94,10 @@ class BufferedIOBase(IOBase):
 
 class FileIO(RawIOBase, BinaryIO):  # type: ignore[misc]  # incompatible definitions of writelines in the base classes
     mode: str
-    name: FileDescriptorOrPath
+    # The type of "name" equals the argument passed in to the constructor,
+    # but that can make FileIO incompatible with other I/O types that assume
+    # "name" is a str. In the future, making FileIO generic might help.
+    name: Any
```
Comment (AlexWaygood) [inline stdlib/io.pyi:100]: Are you sure it's `FileIO.name` that has the incorrect annotation, and not `_WrappedBuffer.name`? Does `_WrappedBuffer.name` _have_ to be a `str`?

Note that `TextIOWrapper.name` just returns the `name` attribute of the underlying buffer, and our annotations for `TextIOWrapper` currently have it just inheriting the `name` property from `typing.TextIO`, meaning that type checkers currently infer `TextIOWrapper.name` to be a property return `str | Any`:

https://github.com/python/typeshed/blob/a3d356c04c15afc22b254afae2e6e7adbbe031a5/stdlib/typing.pyi#L708-L711

So it's a bit weird for us to say "the underlying buffer object has to have a `name` attribute that's a `str`, but the `TextIOWrapper` object could return anything from its `name` property"

Comment (srittau) [reply stdlib/io.pyi:100]: The I/O classes are just hilariously incoherent. I'm fine with just using `Any` in the protocol and/or restoring the `FileIO.name` annotation, although leaving the latter as `Any` for now is probably not a bad idea either.

Comment (AlexWaygood) [reply stdlib/io.pyi:100]: My weak preference would be to use `Any` (or possibly `Incomplete`) for the new protocol for now, as it seems least likely to lead to unexpected and unintended false positives

**Comment by srittau** [comment]
Hrm, cc @JelleZijlstra. Since this PR seem to fix the issues without adding to the primer output. I suggest to merge this first, and then have another look at Jelle's old PR.

**Comment by AlexWaygood** [comment]
> Hrm, cc @JelleZijlstra. Since this PR seem to fix the issues without adding to the primer output. I suggest to merge this first, and then have another look at Jelle's old PR.

That sounds reasonable to me, though if it would be more accurate to have the `read()` method return `ReadableBuffer`, I think it would be safe to make that change. That change should _reduce_ the number of false-positives in user code, if anything.

File: stdlib/io.pyi [hunk 4]
```diff
@@ -146,16 +149,45 @@ class TextIOBase(IOBase):
...
+@type_check_only
+class _WrappedBuffer(Protocol):
+    # "name" is wrapped by TextIOWrapper. Its type is inconsistent between
+    # the various I/O types, see the comments on TextIOWrapper.name and
+    # TextIO.name.
+    @property
+    def name(self) -> Any: ...
+    @property
+    def closed(self) -> bool: ...
+    def read(self, size: int = ..., /) -> ReadableBuffer: ...
```
Comment (JelleZijlstra) [inline stdlib/io.pyi:161]:
```suggestion
    def read(self, size: int, /) -> ReadableBuffer: ...
```
It always [calls read with an argument](https://github.com/python/cpython/blob/81e140d10b77f0a41a5581412e3f3471cc77981f/Modules/_io/textio.c#L1891). Note this change makes the protocol more permissive.

File: stdlib/io.pyi [hunk 5]
```diff
@@ -146,16 +149,45 @@ class TextIOBase(IOBase):
...
+        buffer: _WrappedBuffer,
         encoding: str | None = ...,
         errors: str | None = ...,
         newline: str | None = ...,
         line_buffering: bool = ...,
         write_through: bool = ...,
     ) -> None: ...
+    # Equals the "buffer" argument passed in to the constructor.
+    # TODO: Make TextIOWrapper generic over the buffer argument once
+    # PEP 696 get accepted.
```
Comment (JelleZijlstra) [inline stdlib/io.pyi:190]:
```suggestion
    # PEP 696 gets accepted.
```
Probably need a similar comment for `detach()`.

**Review by JelleZijlstra — COMMENTED** [review]
Thanks for picking this up. The major remaining difference from my PR seems to be that you're not overriding `buffer` and `detach`. This helps to get rid of some of the primer hits.

File: stdlib/io.pyi (hunk 4 above)
Comment (srittau) [reply stdlib/io.pyi:161]: Unfortunately that's not true: https://github.com/python/cpython/blob/81e140d10b77f0a41a5581412e3f3471cc77981f/Modules/_io/textio.c#L1977.

**Review by AlexWaygood — APPROVED** [review]
LGTM. https://github.com/python/typeshed/pull/11420#discussion_r1489781107 is optional.

**Pull request merged**